Adding server for polling client

CuriousVini opened the pull request:
Build: https://builds.cask.co/browse/CDAP-DUT6377-1

CuriousVini committed:
Adding server for polling client

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/handler/RuntimeHandler.java:
```diff
@@ -0,0 +1,145 @@
+ * {@link co.cask.http.HttpHandler} for exposing metadata of a runtime.
+ */
+@Singleton
+@Path(Constants.Gateway.API_VERSION_3 + "/runtime")
```
This is an internal API, should start with v1.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/handler/RuntimeHandler.java:
```diff
@@ -0,0 +1,145 @@
+/**
+ * {@link co.cask.http.HttpHandler} for exposing metadata of a runtime.
+ */
+@Singleton
```
What is this for?

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/handler/RuntimeHandler.java:
```diff
@@ -0,0 +1,145 @@
+  private static final Logger LOG = LoggerFactory.getLogger(RuntimeHandler.class);
+  private static final Gson GSON = new Gson();
+  private static final Type TOKEN_TYPE_SET_STRING = new TypeToken<Set<String>>() {
+  }.getType();
```
Put it on one line, it is easier to read.

`new TypeToken<Set<String>>() { }.getType()`

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/handler/RuntimeHandler.java:
```diff
@@ -0,0 +1,145 @@
+  private static final Gson GSON = new Gson();
+  private static final Type TOKEN_TYPE_SET_STRING = new TypeToken<Set<String>>() {
+  }.getType();
+  private static final Type MAP_STRING_CONSUME_REQUEST_TYPE = new com.google.common.reflect.TypeToken<Map<String,
```
Why use the full class name?

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/handler/RuntimeHandler.java:
```diff
@@ -0,0 +1,145 @@
+    try {
+      String data = request.content().toString(StandardCharsets.UTF_8);
+
+      if (!isValidJSON(data)) {
```
Why need this check? The `GSON.fromJson` would throw exception.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/handler/RuntimeHandler.java:
```diff
@@ -0,0 +1,145 @@
+      }
+
+      responder.sendJson(HttpResponseStatus.OK, GSON.toJson(returnMap));
+    } catch (Exception ex) {
```
What exception would that be? Also, you should use the `HttpExceptionHandler` instead of catching and responding in the handler method.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/handler/RuntimeHandler.java:
```diff
@@ -0,0 +1,145 @@
+  }
+
+  /**
+   * Returns map of topic config -> topic name to be monitored
```
Better to have simple description on what the request and response are.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/handler/RuntimeHandler.java:
```diff
@@ -0,0 +1,145 @@
+          }
+        }
+
+        returnMap.put(topicToFetch.getKey(), list);
```
To have better memory efficiency, better use `JsonWriter` to stream the result back. You can use the `responder.sendChunkStart` to help

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RuntimeServer.java:
```diff
@@ -0,0 +1,93 @@
+  private Cancellable cancelHttpService;
+  private Set<HttpHandler> handlers;
+
+  public RuntimeServer(CConfiguration cConf, InetAddress hostname, int port, Set<HttpHandler> handlers) {
```
Why the handlers are passed in? The runtime server knows what handlers are needed, right?

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RuntimeServer.java:
```diff
@@ -0,0 +1,93 @@
+
+  public RuntimeServer(CConfiguration cConf, InetAddress hostname, int port, Set<HttpHandler> handlers) {
+    this.hostname = hostname;
+    this.port = port;
```
port should comes from `cConf`.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RuntimeServer.java:
```diff
@@ -0,0 +1,93 @@
+  private Cancellable startHttpService(final NettyHttpService httpService) throws Exception {
+    httpService.start();
+
+    return new Cancellable() {
```
Why need to wrap the stop with a `Cancellable`? You can just do the `httpService.stopAndWait` in the shutdown method?

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RuntimeServer.java:
```diff
@@ -0,0 +1,93 @@
+    LoggingContextAccessor.setLoggingContext(new ServiceLoggingContext(NamespaceId.SYSTEM.getNamespace(),
+                                                                       Constants.Logging.COMPONENT_NAME,
+                                                                       Constants.Service.RUNTIME_HTTP));
+    NettyHttpService.Builder httpServiceBuilder = new CommonNettyHttpServiceBuilder(cConf,
```
The server should be running with SSL. Also, can you open a JIRA, if not yet opened, for authenticating the client request is indeed coming from CDAP runtime monitor?

CuriousVini replied:
Created: https://issues.cask.co/browse/CDAP-13252

CuriousVini commented:
@chtyim Thanks for the review, I have addressed comments and as per the discussion, I have modified the protocol which sends a json list of `messages` to the client. I have updated the protocol format https://wiki.cask.co/display/CE/Messaging#Messaging-RESTful

I will modify the client accordingly once server changes are good to go. Please take another pass when you get a chance, thanks!

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/handler/RuntimeHandler.java:
```diff
@@ -0,0 +1,149 @@
+public class RuntimeHandler extends AbstractHttpHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(RuntimeHandler.class);
+  private static final Gson GSON = new Gson();
+  private static final int MAX_MESSAGES_PER_READ = 100;
```
Why needed?

CuriousVini replied:
This is the maximum number of messages stored in memory for server. Basically we send batches of 100 messages as a list.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/handler/RuntimeHandler.java:
```diff
@@ -0,0 +1,149 @@
+  @POST
+  @Path("/metadata")
+  public void metadata(FullHttpRequest request, HttpResponder responder) throws Exception {
+    String data = request.content().toString(StandardCharsets.UTF_8);
```
`data` is super generic. Better call it `requestBody`.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/handler/RuntimeHandler.java:
```diff
@@ -0,0 +1,149 @@
+  public void metadata(FullHttpRequest request, HttpResponder responder) throws Exception {
+    String data = request.content().toString(StandardCharsets.UTF_8);
+
+    Map<String, MonitorConsumeRequest> topicsToFetch = GSON.fromJson(data, MAP_STRING_CONSUME_REQUEST_TYPE);
```
Better call this `consumeRequests`, as the map doesn't have `topic` anymore.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/handler/RuntimeHandler.java:
```diff
@@ -0,0 +1,149 @@
+    ByteBuf buffer = Unpooled.buffer();
+    JsonWriter jsonWriter = new JsonWriter(new OutputStreamWriter(new ByteBufOutputStream(buffer),
+                                                                  StandardCharsets.UTF_8));
+    jsonWriter.beginArray();
```
The result is an array? I thought the result is an object with the keys as the config names from the request body?

CuriousVini replied:
So the result is an array of `TopicMessage` object. If we include `topicConfig` in every batch of 100 messages, then we do not need key value pair.

CuriousVini replied:
yea, the implementation is much easier with topicConfig -> list of messages. I have used this and updated the protocol accordingly in the wiki.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/handler/RuntimeHandler.java:
```diff
@@ -0,0 +1,149 @@
+      List<MonitorMessage> messages = Lists.newArrayListWithCapacity(MAX_MESSAGES_PER_READ);
+
+      // read first batch of messages
+      int eventsRead = readMessages(messages, cConf.get(entry.getKey()), batchSize, lastMessageId);
```
Why need to read messages into at list and then iterate again to json encode one by one? You can just read from the `Iterator` returned from the `MessageFetcher.fetch` call and do the encode on the fly without buffering.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/handler/RuntimeHandler.java:
```diff
@@ -0,0 +1,125 @@
+        Message message = iter.next();
+        GSON.toJson(new MonitorMessage(message.getId(), message.getPayloadAsString(StandardCharsets.UTF_8)),
+                    MonitorMessage.class, jsonWriter);
+        if (buffer.isReadable()) {
```
You are effectively writing out every message in separate chunk, since you are not using `BufferedWriter` for the `JsonWriter`.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/MonitorConsumeRequest.java:
```diff
@@ -23,32 +23,19 @@
-                               boolean inclusive, int limit) {
-    this.topic = topic;
-    this.topicConfig = topicConfig;
+  public MonitorConsumeRequest(@Nullable String messageId, boolean inclusive, int limit) {
```
What is the `inclusive` for? I don't see you are using it at all.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeMonitor.java:
```diff
@@ -93,19 +91,9 @@ private void initializeTopics() throws Exception {
     Set<String> topicsToMonitor = new HashSet<>();
     topicsToMonitor.addAll(Arrays.asList(cConf.get(Constants.RuntimeMonitor.TOPICS).split(",")));
```
These are not `topics` right? Better rename the variable, the constant and the config name.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeMonitor.java:
```diff
@@ -128,8 +118,8 @@ protected void run() {
-          Map<String, List<MonitorMessage>> monitorResponses =
-            GSON.fromJson(response.getResponseBodyAsString(StandardCharsets.UTF_8), MAP_STRING_MESSAGE_TYPE);
+          List<TopicMessage> monitorResponses =
+            GSON.fromJson(response.getResponseBodyAsString(StandardCharsets.UTF_8), LIST_MESSAGE_TYPE);
```
Is this correct? It seems like the response is of type `Map<String, List<MonitorMessage>>`?

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/handler/RuntimeHandler.java:
```diff
@@ -0,0 +1,134 @@
+  @POST
+  @Path("/shutdown")
+  public void shutdown(FullHttpRequest request, HttpResponder responder) throws Exception {
+    shutdownRunnable.run();
```
You need to send the respond back before calling shutdownRunnable.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeMonitor.java:
```diff
@@ -159,11 +160,32 @@ private void publish(String topicConfig, MonitorConsumeRequest request,
+  }
+
+  private void setIsRuntimeInactive(List<MonitorMessage> monitorMessages) {
+    for (MonitorMessage message : monitorMessages) {
```
So this will loop the messages twice?

CuriousVini replied:
yea but since it only loops for programstatus and we  donot expect a lot of messages for program status topic, looping twice will not create big overhead.

CuriousVini replied:
Right, I have handled that by modifying client to throw exception if `monitorResponses` is null

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RuntimeServer.java:
```diff
@@ -0,0 +1,89 @@
+        @Override
+        public void run() {
+          try {
+            shutDown();
```
Should call `stop()` instead of `shutDown`. Otherwise the Service state won't transit correctly.

chtyim replied:
And you can simply just use the method reference `this::stop`.

chtyim commented on cdap-app-fabric/src/test/java/co/cask/cdap/app/runtime/monitor/RuntimeMonitorTest.java:
```diff
@@ -21,195 +21,149 @@
+  public void init() throws IOException, UnsupportedTypeException, TopicAlreadyExistsException {
     cConf = CConfiguration.create();
     cConf.set(Constants.AppFabric.OUTPUT_DIR, System.getProperty("java.io.tmpdir"));
     cConf.set(Constants.AppFabric.TEMP_DIR, System.getProperty("java.io.tmpdir"));
```
Don't use `java.io.tmpdir`. Use `@ClassRule TemporaryFolder` instead.

chtyim replied:
Also, you shouldn't set these two. Set the `Constants.CFG_LOCAL_DATA_DIR` instead.

chtyim commented on cdap-app-fabric/src/test/java/co/cask/cdap/app/runtime/monitor/RuntimeMonitorTest.java:
```diff
@@ -21,195 +21,149 @@
     cConf.set(Constants.AppFabric.TEMP_DIR, System.getProperty("java.io.tmpdir"));
-    cConf.set(Constants.AppFabric.PROGRAM_STATUS_EVENT_TOPIC, "cdap-status");
+    cConf.set(Constants.AppFabric.PROGRAM_STATUS_EVENT_TOPIC, "programStatus");
+    cConf.set(Constants.RuntimeHandler.SERVER_PORT, "1234");
```
Why hardcode the part? This could cause flaky test. You should pass in "0" to have the server bind to an ephemeral port.

chtyim commented on cdap-app-fabric/src/test/java/co/cask/cdap/app/runtime/monitor/RuntimeMonitorTest.java:
```diff
@@ -21,195 +21,149 @@
-    httpServer.start();
+    address = InetAddress.getLoopbackAddress();
+    messagingContext = new MultiThreadMessagingContext(messagingService);
+    messagingService.createTopic(new TopicMetadata(new TopicId("system", "cdap-programStatus")));
```
You shouldn't need to call this if you setup the cdap-default.xml correctly. It should be in the `messaging.system.topics` list. The topic will be created automatically when the messaging service starts.

CuriousVini replied:
This is done to test the case where if a runtime started with latest cdap-site.xml. In that case`runtime` has different topic for program states than cdap runtime monitor.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/handler/RuntimeHandler.java:
```diff
@@ -0,0 +1,135 @@
+    }
+
+    jsonWriter.endObject();
+    jsonWriter.close();
```
Should use `try (JsonWriter jsonWriter = ...) { }` block instead of doing manual close.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/handler/RuntimeHandler.java:
```diff
@@ -0,0 +1,135 @@
+
+  private void writeMessages(JsonWriter jsonWriter, ByteBuf buffer, ChunkResponder chunkResponder, String topic,
+                             int limit, @Nullable String fromMessage) throws TopicNotFoundException, IOException {
+    try (CloseableIterator<Message> iter = messageFetcher.fetch(NamespaceId.SYSTEM.getNamespace(), topic, limit,
```
So if there is exception, then the resulting json object would be broken?

CuriousVini replied:
In that case the response will be a string: https://github.com/caskdata/cdap/blob/develop/cdap-common/src/main/java/co/cask/cdap/common/HttpExceptionHandler.java#L72

The Client should handle this case. I will modify the client accordingly.

chtyim replied:
It won't work.. You already sent the partial json, how can it be a string then?

CuriousVini replied:
Right, I have handled that by modifying client to throw exception if `monitorResponses` is null

chtyim replied:
How is it on the client side?? Imagine that you have outputted this so far:

```
{
  "program.state.topic" : [
    { "messageId" : "xyz", "message" : "RUNNING" },
```

Then on the next message in the loop, you got an exception. If you just let the exception propagate, then the client will get the broken json with a 200 response code. Do you expect the client to catch that?

I think it's better to handle it on the server side to always have a valid json response. If there is exception, you can catch it and close the json object appropriately..

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeMonitor.java:
```diff
@@ -90,22 +93,12 @@ protected void startUp() throws Exception {
-    Map<String, String> configToTopic = GSON.fromJson(response.getResponseBodyAsString(StandardCharsets.UTF_8),
-                                                      MAP_STRING_STRING_TYPE);
+    Set<String> topicsConfigsToMonitor = new HashSet<>();
+    topicsConfigsToMonitor.addAll(Arrays.asList(cConf.get(Constants.RuntimeMonitor.TOPICS_CONFIGS).split(",")));
```
The metrics could have multiple topics. You might want to refer to MessagingServiceUtils.getSystemTopics on how it is parsed (you could move that method to common if needed).

CuriousVini replied:
Since we are using topicConfig per topic and keep track of offsets for each topic in RuntimeMonitor, this would require a new rest endpoint (because Runtime can be using different number of metrics than RuntimeMonitor). I will handle that in my next PR when I am adding `kill` endpoint

chtyim replied:
ok. :)

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RuntimeServer.java:
```diff
@@ -0,0 +1,89 @@
+        public void run() {
+          try {
+            stop();
+          } catch (Exception e) {
```
I don't think `stop()` throws exception. You could simplify this using method reference `new RuntimeHandler(cConf, messageFetcher, this::stop);`

CuriousVini replied:
fixed.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/handler/RuntimeHandler.java:
```diff
@@ -0,0 +1,136 @@
+                        entry.getValue().getMessageId());
+        } finally {
+          jsonWriter.endArray();
+          jsonWriter.endObject();
```
This doesn't match with the `beginObject` scope, right? The `beginObject` only called once per request in line 91, but the `endObject` is called once per loop.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/handler/RuntimeHandler.java:
```diff
@@ -0,0 +1,136 @@
+        jsonWriter.beginArray();
+
+        try {
+          writeMessages(jsonWriter, buffer, chunkResponder, cConf.get(entry.getKey()), entry.getValue().getLimit(),
```
You have to make sure writeMessages doesn't throw, right? Otherwise there could be unsend data in the `buffer`.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeMonitor.java:
```diff
@@ -71,6 +73,7 @@
   private final long pollTimeMillis;
   private volatile Thread runThread;
   private volatile boolean stopped;
+  private volatile boolean isRuntimeInactive;
```
I don't think you need volatile. Also, a better name is `programFinished`, because the runtime is still active, otherwise you can't even talk to it. Moreover, boolean variable shouldn't start with `is`, because that's the name prefix for the boolean getter (e.g. a `boolean running` would have getting method `isRunning`).

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeMonitor.java:
```diff
@@ -143,14 +138,20 @@ protected void run() {
-  private void publish(String topicConfig, MonitorConsumeRequest request,
-                       List<MonitorMessage> messages) throws Exception {
+  private void publish(String topicConfig, List<MonitorMessage> messages)
+    throws Exception {
```
no need to wrap the line.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeMonitor.java:
```diff
@@ -159,11 +160,32 @@ private void publish(String topicConfig, MonitorConsumeRequest request,
+  private void setIsRuntimeInactive(List<MonitorMessage> monitorMessages) {
+    for (MonitorMessage message : monitorMessages) {
+      Notification notification = GSON.fromJson(message.getMessage(), Notification.class);
+      String programStatus = notification.getProperties().get(ProgramOptionConstants.PROGRAM_STATUS);
```
Please also open a JIRA / add to the existing security JIRA that the runtime monitor should check that the messages is actually intended for the program run ID that it is monitoring.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RuntimeServer.java:
```diff
@@ -0,0 +1,80 @@
+
+  @Override
+  protected void shutDown() throws Exception {
+    httpService.stop();
```
You have to call `stopAndWait` for this one.

CuriousVini replied:
`NettyHttpService` does not have that method.

chtyim replied:
oh, right. Sorry my bad :)

chtyim reviewed (approved):
LGTM. Please squash before merging.

CuriousVini commented:
Build: https://builds.cask.co/browse/CDAP-DUT6377-6

CuriousVini commented:
Build is green. Squashing and merging.
